scipy/optimize/_minpack_py.py

```diff
@@ -465,11 +466,24 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
     if full_output:
         cov_x = None
         if info in LEASTSQ_SUCCESS:
-            perm = take(eye(n), retval[1]['ipvt'] - 1, 0)
+            # This was
+            # perm = take(eye(n), retval[1]['ipvt'] - 1, 0)
+            # r = triu(transpose(retval[1]['fjac'])[:n, :])
+            # R = dot(r, perm)
+            # cov_x = inv(dot(transpose(R), R))
+            # but the explicit dot product was not necessary and sometimes
+            # the result was not symmetric positive definite. See gh-4555.
+            perm = retval[1]['ipvt'] - 1
+            n = len(perm)
             r = triu(transpose(retval[1]['fjac'])[:n, :])
-            R = dot(r, perm)
+            inv_triu = linalg.get_lapack_funcs('trtri', (r,))
             try:
-                cov_x = inv(dot(transpose(R), R))
+                # inverse of permuted matrix is a permuation of matrix inverse
+                invR, trtri_info = inv_triu(r)  # default: upper, non-unit diag
+                if trtri_info:
+                    raise LinAlgError(f'trtri returned info {trtri_info}')
+                invR[perm] = invR[np.arange(n)]
```

Comment:
isn't this basically `invR[perm] = invR` since you basically get all rows in the natural ordering? Might have missed the point though.

Reply:
Yes, probably. I think at one point when I was playing around I was permuting the columns instead of the rows and I needed the `arange`. I'll test it without them.

Reply:
Ah, of course: a copy of some sort is needed.  

```python3
import numpy as np
A = np.eye(3)
perm = [1, 2, 0]
A[perm] = A
# array([[1., 0., 0.],
#        [1., 0., 0.],
#        [1., 0., 0.]])
```
vs
```python3
A = np.eye(3)
A[perm] = A.copy()
# array([[0., 0., 1.],
#        [1., 0., 0.],
#        [0., 1., 0.]])
```